src/psyclone/domain/nemo/transformations/nemo_arrayrange2loop_trans.py

```diff
@@ -0,0 +1,275 @@
+            if the Range node is not the outermost Range node of the \
+            ArrayReference or if that ArrayReference does not \
+            consitute the left hand side of an Assignment node.
+
+        '''
+        # Am I Range node?
+        if not isinstance(node, Range):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be a PSyIR Range, but "
+                "found '{0}'.".format(type(node).__name__))
+        # Am I within an array reference?
+        if not node.parent or not isinstance(node.parent, ArrayReference):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node, but found '{0}'.".format(type(node.parent).__name__))
+        array_ref = node.parent
+        # Is the array reference within an assignment?
+        if not array_ref.parent or not isinstance(array_ref.parent,
+                                                  Assignment):
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within an Assignment node, but found '{0}'."
+                .format(type(array_ref.parent).__name__))
+        assignment = array_ref.parent
+        # Is the array reference the lhs of the assignment?
+        if assignment.lhs is not array_ref:
+            raise TransformationError(
+                "Error in NemoArrayRange2LoopTrans transformation. The "
+                "supplied node argument should be within an ArrayReference "
+                "node that is within the left-hand-side of an Assignment "
+                "node, but it is on the right-hand-side.")
+        # Is the Range node the outermost Range (as if not, the
+        # transformation would be invalid).
+        my_index = node.parent.children.index(node)
+        num_siblings = len(node.parent.children)
+        for idx in range(my_index+1, num_siblings):
+            following_node = node.parent.children[idx]
+            if isinstance(following_node, Range):
+                raise TransformationError(
+                    "Error in NemoArrayRange2LoopTrans transformation. This "
+                    "transformation can only be applied to the outermost "
+                    "Range.")
+
+
+def _get_outer_index(array):
+    '''Find the outermost index of the array that is a Range node. If one
+    does not exist then raise an exception.
+
+    :param array: the array being examined.
+    :type array: :py:class:`psyclone.psyir.nodes.ArrayReference`
+
+    :returns: the outermost index of the array that is a Range node.
+
+    :raises IndexError: if the array does not contain a Range node.
+
+    '''
+    idx = len(array.children)-1
```

Comment:
I think this could be nicer, e.g:

    for child in reversed(array.children):
        if isinstance(child, Range):
            break
    else:
        raise IndexError
    return child.position

Reply:
Yes, much better. I changed your proposed code to `return child.position` instead of `break` which allowed me to remove the else and just `raise IndexError` after the loop.